node/libs/concurrency/src/sync/prunable_mpsc/tests.rs

```diff
@@ -0,0 +1,94 @@
+#[tokio::test]
+async fn test_prunable_mpsc() {
+    let ctx = ctx::test_root(&ctx::RealClock);
+
+    #[derive(Debug, Clone)]
+    struct ValueType(usize, usize);
+
+    #[allow(clippy::type_complexity)]
+    let (send, mut recv): (
+        super::Sender<ValueType, Result<(), usize>>,
+        super::Receiver<ValueType, Result<(), usize>>,
+    ) = super::channel(|a: &ValueType, b: &ValueType| {
+        // Prune values with the same i.
+        a.1 == b.1
+    });
+
+    let res: Result<(), ctx::Canceled> = crate::scope::run!(&ctx, |ctx, s| async move {
+        // Pre-send sets 0 and 1, 1000 values each.
+        // Set 0 is expected to be pruned and dropped.
+        let values = (0..2000).map(|i| ValueType(i / 1000, i % 1000));
+        for val in values {
+            let res_recv = send.send(val.clone()).await;
+            s.spawn(async move {
+                let res = res_recv.recv_or_disconnected(ctx).await;
+                match val.0 {
+                    // set 0 values are expected to be pruned and dropped.
+                    0 => assert_matches!(res, Ok(Err(crate::sync::Disconnected))),
+                    // set 1 values are expected to return `Ok(())`.
+                    1 => assert_matches!(res, Ok(Ok(Ok(())))),
+                    _ => unreachable!(),
+                }
+                Ok(())
+            });
+        }
+        // Send set 2.
+        s.spawn(async move {
+            let values = (1000..2000).map(|i| ValueType(2, i));
+            for val in values {
+                let res_recv = send.send(val.clone()).await;
+                s.spawn(async move {
+                    let res = res_recv.recv_or_disconnected(ctx).await;
+                    let i = val.1;
+                    match val.0 {
+                        // set 2 values are expected to return `Err(i)`.
+                        2 => assert_matches!(res, Ok(Ok(Err(err))) => {
+                            assert_eq!(err, i);
+                        }),
+                        _ => unreachable!(),
+                    };
+                    Ok(())
+                });
+            }
+            Ok(())
+        });
+        // Receive.
+        s.spawn(async move {
+            let mut i = 0;
+            loop {
+                let (val, res_send) = recv.recv(ctx).await.unwrap();
+                assert_eq!(val.1, i);
+                match val.0 {
+                    // set 0 is expected to be pruned and dropped.
+                    0 => unreachable!(),
+                    // Return `Ok(())` for set 1.
+                    1 => res_send.send(Ok(())).unwrap(),
+                    // Return `Err(i)` for set 2.
+                    2 => res_send.send(Err(i)).unwrap(),
+                    _ => unreachable!(),
+                };
+                i += 1;
+                if i == 2000 {
+                    assert!(
+                        timeout(Duration::from_secs(0), recv.recv(ctx))
```

Comment:
nit: as previously, `timeout()` doesn't guarantee the expected semantics here.

Reply:
Is there another way you'd recommend doing this, while only having `recv()`?

Reply:
I'd suggest inspecting directly the channel state. But it wouldn't hurt to have a public try_recv() and/or empty() methods.

Reply:
This would test the buffer's state, but not the waiting functionality.

Reply:
This would test the channel's state, but not the waiting functionality.

Reply:
well, recv() signature prevents it from returning without providing an element of T, which is not required to have any constructor, so you might say that a test is not needed. But if you really want to have a test of that you can add a small realtime timeout, i.e. call `recv.recv(&ctx.with_timeout(time::Duration::milliseconds(10))).await`

Reply:
Fixed, changes here: https://github.com/matter-labs/era-consensus/pull/56/commits/ab5f4cb87a3bb95b024f5a9f80a0f6e43e2c43f9